src/vring.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+// Copyright 2021 Alibaba Cloud Computing. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Struct to maintain state information and manipulate vhost-user queues.
+
+use std::fs::File;
+use std::io;
+use std::os::unix::io::{FromRawFd, IntoRawFd};
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use virtio_queue::Queue;
+use vm_memory::{GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+use vmm_sys_util::eventfd::EventFd;
+
+/// Struct to maintain raw state information for a vhost-user queue.
+pub struct VringState<M: GuestAddressSpace> {
+    queue: Queue<M>,
+    kick: Option<EventFd>,
+    call: Option<EventFd>,
+    err: Option<EventFd>,
+    enabled: bool,
+}
+
+impl<M: GuestAddressSpace> VringState<M> {
+    fn new(mem: M, max_queue_size: u16) -> Self {
+        VringState {
+            queue: Queue::new(mem, max_queue_size),
+            kick: None,
+            call: None,
+            err: None,
+            enabled: false,
+        }
+    }
+
+    /// Get a mutable reference to the underlying raw `Queue` object.
+    pub fn get_queue_mut(&mut self) -> &mut Queue<M> {
+        &mut self.queue
+    }
+
+    /// Get a immutable reference to the kick event fd.
+    pub fn get_kick(&self) -> &Option<EventFd> {
+        &self.kick
+    }
+}
+
+/// Struct to maintain state information and manipulate a vhost-user queue.
+#[derive(Clone)]
+pub struct Vring<M: GuestAddressSpace = GuestMemoryAtomic<GuestMemoryMmap>> {
+    state: Arc<RwLock<VringState<M>>>,
+}
+
+impl<M: GuestAddressSpace> Vring<M> {
+    /// Get a immutable guard to the underlying raw `VringState` object.
+    pub fn get_ref(&self) -> RwLockReadGuard<VringState<M>> {
+        self.state.read().unwrap()
+    }
+
+    /// Get a mutable guard to the underlying raw `VringState` object.
+    pub fn get_mut(&self) -> RwLockWriteGuard<VringState<M>> {
+        self.state.write().unwrap()
+    }
+
+    /// Add an used descriptor into the used queue.
+    pub fn add_used(&self, mem: &M::M, desc_index: u16, len: u32) -> Option<u16> {
+        self.get_mut()
+            .get_queue_mut()
+            .add_used(mem, desc_index, len)
```

Comment:
The current signature for `virtio_queue::Queue::add_used` is like this:

```
 pub fn add_used(&mut self, head_index: u16, len: u32) -> Result<(), Error>
````

Does this PR require some pending change to `virtio_queue`?

Reply:
yes, I do have some pending PRs for vm-virtio, all our existing projects are using signature of 
```
pub fn add_used(&self, mem: &M::M, desc_index: u16, len: u32) -> Option<u16>
```
I will change the signature to follow current interface of vm-virtio/virtio-queue.